src/main/java/com/avispl/dal/communicator/polycom/videoos/PolycomVideoOS.java

```diff
@@ -0,0 +1,1046 @@
+import org.springframework.http.client.ClientHttpRequestInterceptor;
+import org.springframework.http.client.ClientHttpResponse;
+import org.springframework.util.CollectionUtils;
+import org.springframework.web.client.RestTemplate;
+
+import javax.security.auth.login.FailedLoginException;
+import java.io.IOException;
+import java.util.*;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * PolycomVideoOS is a communicator class for X30/X50/G7500 devices.
+ * The communicator is based on RestCommunicator class and is using VideoOS REST API to communicate with devices.
+ * In order to have codec specific features available, CallController interface is used with dial(), hangup()
+ * and callStatus() methods implemented.
+ * sendMessage() is not implemented since there's no such functionality available in VideoOS REST API for now.
+ */
+public class PolycomVideoOS extends RestCommunicator implements CallController, Monitorable, Controller {
+
+    private ClientHttpRequestInterceptor videoOSInterceptor = new PolycomVideoOSInterceptor();
+
+    /**
+     * Interceptor for RestTemplate that injects
+     * Cookie header and reacts to the communicator state - after reboot the certificates on the device get
+     * refreshed that results with timeouts on https requests. In order to overcome this - communicator is being internally
+     * destroyed on the successful reboot action. Hovewer, we cannot guarantee that any other resource does not issue
+     * reboot or certificates do not get refreshed for other reasons, so we check for a specific error on
+     * authorization event and reinitialize the communicator here.
+     */
+    class PolycomVideoOSInterceptor implements ClientHttpRequestInterceptor {
+        @Override
+        public ClientHttpResponse intercept(HttpRequest request, byte[] body, ClientHttpRequestExecution execution) throws IOException {
+
+            ClientHttpResponse response = execution.execute(request, body);
+            if (response.getRawStatusCode() == 403) {
+                try {
+                    authenticate();
+                    response = execution.execute(request, body);
+                } catch (Exception e) {
+                    logger.error("Authentication failed during interception: " + e.getMessage());
+                    if (e.getMessage().startsWith("Cannot reach resource") && e.getMessage().contains(SESSION)) {
+                        // In case it's been rebooted by some other resource - we need to react to that.
+                        // Normally we expect that the authorization request timeouts, since previous one failed with
+                        // a specific error code, so basically we do the same as before - authenticate + retry previous
+                        // request but with the destroy action called first, since init will be done next when
+                        // authentication is requested.
+                        internalDestroy();

```

Comment:
Must not destroy itself, we already talked about it. Life cycle is managed by whoever is consuming the adapter, not by adapter itself. Whatever you call destroy here fore (which is not obvious) can (and should) be achieved in another way.